pkg/client/build.go

```diff
@@ -798,35 +790,125 @@ func (c *Client) processBuildpacks(ctx context.Context, builderImage imgutil.Ima
 
 				order = newOrder
 			}
-		case buildpack.IDLocator:
-			id, version := buildpack.ParseIDLocator(bp)
-			order = appendBuildpackToOrder(order, dist.ModuleInfo{
-				ID:      id,
-				Version: version,
-			})
 		default:
-			imageOS, err := builderImage.OS()
+			newFetchedBPs, moduleInfo, err := c.fetchBuildpack(ctx, bp, relativeBaseDir, builderImage, builderBPs, opts)
 			if err != nil {
-				return fetchedBPs, order, errors.Wrapf(err, "getting OS from %s", style.Symbol(builderImage.Name()))
+				return fetchedBPs, order, err
 			}
-			mainBP, depBPs, err := c.buildpackDownloader.Download(ctx, bp, buildpack.DownloadOptions{
-				RegistryName:    registry,
-				ImageOS:         imageOS,
-				RelativeBaseDir: relativeBaseDir,
-				Daemon:          !publish,
-				PullPolicy:      pullPolicy,
-			})
-			if err != nil {
-				return fetchedBPs, order, errors.Wrap(err, "downloading buildpack")
+			fetchedBPs = append(fetchedBPs, newFetchedBPs...)
+			order = appendBuildpackToOrder(order, *moduleInfo)
+		}
+	}
+
+	if (len(order) == 0 || len(order[0].Group) == 0) && len(builderOrder) > 0 {
+		preBuildpacks := opts.PreBuildpacks
+		postBuildpacks := opts.PostBuildpacks
+		if len(preBuildpacks) == 0 && len(opts.ProjectDescriptor.Build.Pre.Buildpacks) > 0 {
```

Comment:
i'm almost surely just missing something but it's not obvious to me why the `len(preBuildpacks)==0` check is important -- we wouldn't want to copy in the buildpacks from the projectdescriptor regardless?

Reply:
hmm, that's a good question... if we should merge `--pre-buildpacks` with `opts.ProjectDescriptor.Build.Pre.Buildpacks`. Or have it override (as it's written now.

Reply:
I'm leaning toward _override_ because i'm not sure it's obvious which would come first in the order

Reply:
I think merging stuff gets messy, so I'm :+1: on overriding.

Reply:
cool cool, thanks :)